{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LTEE-FBA-Jenga.ipynb by Rohan Maddamsetti\n",
    "\n",
    "See documentation at: \n",
    "https://cobrapy.readthedocs.io/en/latest/  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib as plt\n",
    "import cobra.test\n",
    "import cobra\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from os import path\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## use Gurobi. Necessary for constructing minimal genomes!\n",
    "cobra_config = cobra.Configuration()\n",
    "cobra_config.solver = \"gurobi\"\n",
    "cobra_config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_dict_of_KO_strains(anc_model, gene_vec=[]):\n",
    "    deletion_cobra_models = {}\n",
    "    \n",
    "    gene_str_list = [str(gene_obj) for gene_obj in anc_model.genes]\n",
    "    if len(gene_vec): ## then filter for the genes (strings) in gene_vec.\n",
    "        gene_str_list = [x for x in gene_str_list if x in gene_vec]\n",
    "        \n",
    "    for genestr in gene_str_list:\n",
    "        cur_KO_model = anc_model.copy()\n",
    "        cur_KO_model.id = '_'.join([anc_model.id, genestr, \"knockout\"])\n",
    "        cur_gene = cur_KO_model.genes.get_by_id(genestr)\n",
    "        cobra.manipulation.delete_model_genes(cur_KO_model, cur_gene)\n",
    "        deletion_cobra_models[genestr] = cur_KO_model\n",
    "    return deletion_cobra_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_LTEE_clone_cobra_models(KOed_genes_df, basic_model, using_locus_tag=False):\n",
    "    \n",
    "    LTEE_cobra_models = {}\n",
    "    nonmutator_pops = [\"Ara-5\", \"Ara-6\", \"Ara+1\", \"Ara+2\", \"Ara+4\", \"Ara+5\"]\n",
    "    mutator_pops = [\"Ara-1\", \"Ara-2\", \"Ara-3\", \"Ara-4\", \"Ara+3\", \"Ara+6\"]\n",
    "    LTEE_pop_vec = nonmutator_pops + mutator_pops\n",
    "    model_genes = [x.id for x in basic_model.genes]\n",
    "    \n",
    "    for LTEE_pop in LTEE_pop_vec:\n",
    "        cur_KO_model = basic_model.copy()\n",
    "        cur_KO_model.id = LTEE_pop + \"_50K_A_clone\"\n",
    "        is_cur_pop = (KOed_genes_df[\"Population\"] == LTEE_pop)\n",
    "        \n",
    "        if using_locus_tag:\n",
    "            KOed_genes = [x for x in KOed_genes_df[is_cur_pop].locus_tag]\n",
    "        else:\n",
    "            KOed_genes = [x for x in KOed_genes_df[is_cur_pop].blattner]\n",
    "        \n",
    "        genes_to_remove = [cur_KO_model.genes.get_by_id(x) for x in KOed_genes if x in model_genes]\n",
    "        if genes_to_remove:\n",
    "            cobra.manipulation.delete_model_genes(cur_KO_model, genes_to_remove)\n",
    "        \n",
    "        LTEE_cobra_models[LTEE_pop] = cur_KO_model\n",
    "    \n",
    "    return LTEE_cobra_models"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question: have any of the BiGG core, superessential, or specialist/generalist enzymes been knocked out in the 50K LTEE A clones?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "KOed_50K_metabolic_genes = pd.read_csv(\"../results/metabolic-enzymes/KOed-metabolic-enzymes-in-LTEE-50K-A-clones.csv\")\n",
    "\n",
    "KOed_BiGG_core = KOed_50K_metabolic_genes[KOed_50K_metabolic_genes['MetabolicClass'] == \"BiGG_core\"]\n",
    "KOed_superessential = KOed_50K_metabolic_genes[KOed_50K_metabolic_genes['MetabolicClass'] == \"Superessential\"]\n",
    "KOed_specialist = KOed_50K_metabolic_genes[KOed_50K_metabolic_genes['MetabolicClass'] == \"Specialist\"]\n",
    "KOed_generalist = KOed_50K_metabolic_genes[KOed_50K_metabolic_genes['MetabolicClass'] == \"Generalist\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Gene</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>blattner</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>product</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>MetabolicClass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>appB</td>\n",
       "      <td>ECB_00982</td>\n",
       "      <td>b0979</td>\n",
       "      <td>1137</td>\n",
       "      <td>cytochrome bd-II oxidase subunit II</td>\n",
       "      <td>1056387</td>\n",
       "      <td>1057523</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>pykF</td>\n",
       "      <td>ECB_01645</td>\n",
       "      <td>b1676</td>\n",
       "      <td>1413</td>\n",
       "      <td>pyruvate kinase</td>\n",
       "      <td>1732965</td>\n",
       "      <td>1734377</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>gltB</td>\n",
       "      <td>ECB_03077</td>\n",
       "      <td>b3212</td>\n",
       "      <td>4554</td>\n",
       "      <td>glutamate synthase large subunit</td>\n",
       "      <td>3289867</td>\n",
       "      <td>3294420</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>pflC</td>\n",
       "      <td>ECB_03837</td>\n",
       "      <td>b3952</td>\n",
       "      <td>879</td>\n",
       "      <td>pyruvate formate lyase II activase</td>\n",
       "      <td>4126365</td>\n",
       "      <td>4127243</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>pfkB</td>\n",
       "      <td>ECB_01692</td>\n",
       "      <td>b1723</td>\n",
       "      <td>930</td>\n",
       "      <td>6-phosphofructokinase II</td>\n",
       "      <td>1783635</td>\n",
       "      <td>1784564</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>fbaB</td>\n",
       "      <td>ECB_02025</td>\n",
       "      <td>b2097</td>\n",
       "      <td>1053</td>\n",
       "      <td>fructose-bisphosphate aldolase</td>\n",
       "      <td>2131641</td>\n",
       "      <td>2132693</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>dld</td>\n",
       "      <td>ECB_02063</td>\n",
       "      <td>b2133</td>\n",
       "      <td>1716</td>\n",
       "      <td>D-lactate dehydrogenase FAD-binding NADH indep...</td>\n",
       "      <td>2173423</td>\n",
       "      <td>2175138</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>eutI</td>\n",
       "      <td>ECB_02349</td>\n",
       "      <td>b2458</td>\n",
       "      <td>1017</td>\n",
       "      <td>predicted phosphotransacetylase subunit</td>\n",
       "      <td>2493318</td>\n",
       "      <td>2494334</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Ara-3</td>\n",
       "      <td>kgtP</td>\n",
       "      <td>ECB_02481</td>\n",
       "      <td>b2587</td>\n",
       "      <td>1299</td>\n",
       "      <td>alpha-ketoglutarate transporter</td>\n",
       "      <td>2645941</td>\n",
       "      <td>2647239</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Ara-3</td>\n",
       "      <td>tdcE</td>\n",
       "      <td>ECB_02981</td>\n",
       "      <td>b3114</td>\n",
       "      <td>2295</td>\n",
       "      <td>pyruvate formate-lyase 4/2-ketobutyrate format...</td>\n",
       "      <td>3193158</td>\n",
       "      <td>3195452</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Ara-3</td>\n",
       "      <td>tdcD</td>\n",
       "      <td>ECB_02982</td>\n",
       "      <td>b3115</td>\n",
       "      <td>1221</td>\n",
       "      <td>propionate kinase/acetate kinase C anaerobic</td>\n",
       "      <td>3195486</td>\n",
       "      <td>3196706</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Ara-3</td>\n",
       "      <td>aceA</td>\n",
       "      <td>ECB_03887</td>\n",
       "      <td>b4015</td>\n",
       "      <td>1305</td>\n",
       "      <td>isocitrate lyase</td>\n",
       "      <td>4196554</td>\n",
       "      <td>4197858</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Ara-3</td>\n",
       "      <td>rpiB</td>\n",
       "      <td>ECB_03962</td>\n",
       "      <td>b4090</td>\n",
       "      <td>450</td>\n",
       "      <td>ribose-5-phosphate isomerase B</td>\n",
       "      <td>4292172</td>\n",
       "      <td>4292621</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Ara-3</td>\n",
       "      <td>fumB</td>\n",
       "      <td>ECB_03993</td>\n",
       "      <td>b4122</td>\n",
       "      <td>1647</td>\n",
       "      <td>anaerobic class I fumarate hydratase fumarase B</td>\n",
       "      <td>4324395</td>\n",
       "      <td>4326041</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>pflB</td>\n",
       "      <td>ECB_00907</td>\n",
       "      <td>b0903</td>\n",
       "      <td>2283</td>\n",
       "      <td>pyruvate formate lyase I</td>\n",
       "      <td>968365</td>\n",
       "      <td>970647</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>acnA</td>\n",
       "      <td>ECB_01253</td>\n",
       "      <td>b1276</td>\n",
       "      <td>2676</td>\n",
       "      <td>aconitate hydratase</td>\n",
       "      <td>1334203</td>\n",
       "      <td>1336878</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>sfcA</td>\n",
       "      <td>ECB_01437</td>\n",
       "      <td>b1479</td>\n",
       "      <td>1698</td>\n",
       "      <td>malate dehydrogenase decarboxylating NAD-requi...</td>\n",
       "      <td>1527910</td>\n",
       "      <td>1529607</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>fumC</td>\n",
       "      <td>ECB_01580</td>\n",
       "      <td>b1611</td>\n",
       "      <td>1404</td>\n",
       "      <td>fumarate hydratase</td>\n",
       "      <td>1662380</td>\n",
       "      <td>1663783</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>fbaB</td>\n",
       "      <td>ECB_02025</td>\n",
       "      <td>b2097</td>\n",
       "      <td>1053</td>\n",
       "      <td>fructose-bisphosphate aldolase</td>\n",
       "      <td>2131641</td>\n",
       "      <td>2132693</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>yghK</td>\n",
       "      <td>ECB_02843</td>\n",
       "      <td>b2975</td>\n",
       "      <td>1683</td>\n",
       "      <td>glycolate transporter</td>\n",
       "      <td>3046489</td>\n",
       "      <td>3048171</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>glcB</td>\n",
       "      <td>ECB_02844</td>\n",
       "      <td>b2976</td>\n",
       "      <td>2172</td>\n",
       "      <td>malate synthase</td>\n",
       "      <td>3048526</td>\n",
       "      <td>3050697</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>pitB</td>\n",
       "      <td>ECB_02863</td>\n",
       "      <td>b2987</td>\n",
       "      <td>1500</td>\n",
       "      <td>phosphate transporter</td>\n",
       "      <td>3068959</td>\n",
       "      <td>3070458</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>maeB</td>\n",
       "      <td>ECB_02354</td>\n",
       "      <td>b2463</td>\n",
       "      <td>2280</td>\n",
       "      <td>malic enzyme</td>\n",
       "      <td>2496926</td>\n",
       "      <td>2499205</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>gltB</td>\n",
       "      <td>ECB_03077</td>\n",
       "      <td>b3212</td>\n",
       "      <td>4554</td>\n",
       "      <td>glutamate synthase large subunit</td>\n",
       "      <td>3289867</td>\n",
       "      <td>3294420</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Ara-6</td>\n",
       "      <td>gltD</td>\n",
       "      <td>ECB_03078</td>\n",
       "      <td>b3213</td>\n",
       "      <td>1419</td>\n",
       "      <td>glutamate synthase 4Fe-4S protein small subunit</td>\n",
       "      <td>3294433</td>\n",
       "      <td>3295851</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>pflB</td>\n",
       "      <td>ECB_00907</td>\n",
       "      <td>b0903</td>\n",
       "      <td>2283</td>\n",
       "      <td>pyruvate formate lyase I</td>\n",
       "      <td>968365</td>\n",
       "      <td>970647</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>yghK</td>\n",
       "      <td>ECB_02843</td>\n",
       "      <td>b2975</td>\n",
       "      <td>1683</td>\n",
       "      <td>glycolate transporter</td>\n",
       "      <td>3046489</td>\n",
       "      <td>3048171</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Ara+2</td>\n",
       "      <td>appC</td>\n",
       "      <td>ECB_00981</td>\n",
       "      <td>b0978</td>\n",
       "      <td>1545</td>\n",
       "      <td>cytochrome bd-II oxidase subunit I</td>\n",
       "      <td>1054831</td>\n",
       "      <td>1056375</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Ara+2</td>\n",
       "      <td>tktB</td>\n",
       "      <td>ECB_02356</td>\n",
       "      <td>b2465</td>\n",
       "      <td>2004</td>\n",
       "      <td>transketolase 2 thiamin-binding</td>\n",
       "      <td>2500464</td>\n",
       "      <td>2502467</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Ara+2</td>\n",
       "      <td>sgcE</td>\n",
       "      <td>ECB_04166</td>\n",
       "      <td>b4301</td>\n",
       "      <td>633</td>\n",
       "      <td>ribulose-phosphate 3-epimerase</td>\n",
       "      <td>4507043</td>\n",
       "      <td>4507675</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Ara+3</td>\n",
       "      <td>pflB</td>\n",
       "      <td>ECB_00907</td>\n",
       "      <td>b0903</td>\n",
       "      <td>2283</td>\n",
       "      <td>pyruvate formate lyase I</td>\n",
       "      <td>968365</td>\n",
       "      <td>970647</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Ara+3</td>\n",
       "      <td>adhE</td>\n",
       "      <td>ECB_01215</td>\n",
       "      <td>b1241</td>\n",
       "      <td>2676</td>\n",
       "      <td>fused acetaldehyde-CoA dehydrogenase/iron-depe...</td>\n",
       "      <td>1294239</td>\n",
       "      <td>1296914</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Ara+3</td>\n",
       "      <td>focB</td>\n",
       "      <td>ECB_02384</td>\n",
       "      <td>b2492</td>\n",
       "      <td>849</td>\n",
       "      <td>predicted formate transporter</td>\n",
       "      <td>2534754</td>\n",
       "      <td>2535602</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Ara+4</td>\n",
       "      <td>pykF</td>\n",
       "      <td>ECB_01645</td>\n",
       "      <td>b1676</td>\n",
       "      <td>1413</td>\n",
       "      <td>pyruvate kinase</td>\n",
       "      <td>1732965</td>\n",
       "      <td>1734377</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>Ara+4</td>\n",
       "      <td>fbaB</td>\n",
       "      <td>ECB_02025</td>\n",
       "      <td>b2097</td>\n",
       "      <td>1053</td>\n",
       "      <td>fructose-bisphosphate aldolase</td>\n",
       "      <td>2131641</td>\n",
       "      <td>2132693</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>Ara+4</td>\n",
       "      <td>gltB</td>\n",
       "      <td>ECB_03077</td>\n",
       "      <td>b3212</td>\n",
       "      <td>4554</td>\n",
       "      <td>glutamate synthase large subunit</td>\n",
       "      <td>3289867</td>\n",
       "      <td>3294420</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>yghK</td>\n",
       "      <td>ECB_02843</td>\n",
       "      <td>b2975</td>\n",
       "      <td>1683</td>\n",
       "      <td>glycolate transporter</td>\n",
       "      <td>3046489</td>\n",
       "      <td>3048171</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>glcB</td>\n",
       "      <td>ECB_02844</td>\n",
       "      <td>b2976</td>\n",
       "      <td>2172</td>\n",
       "      <td>malate synthase</td>\n",
       "      <td>3048526</td>\n",
       "      <td>3050697</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>pitB</td>\n",
       "      <td>ECB_02863</td>\n",
       "      <td>b2987</td>\n",
       "      <td>1500</td>\n",
       "      <td>phosphate transporter</td>\n",
       "      <td>3068959</td>\n",
       "      <td>3070458</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>Ara+5</td>\n",
       "      <td>gltP</td>\n",
       "      <td>ECB_03949</td>\n",
       "      <td>b4077</td>\n",
       "      <td>1314</td>\n",
       "      <td>glutamate/aspartate:proton symporter</td>\n",
       "      <td>4273414</td>\n",
       "      <td>4274727</td>\n",
       "      <td>1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>yghK</td>\n",
       "      <td>ECB_02843</td>\n",
       "      <td>b2975</td>\n",
       "      <td>1683</td>\n",
       "      <td>glycolate transporter</td>\n",
       "      <td>3046489</td>\n",
       "      <td>3048171</td>\n",
       "      <td>-1</td>\n",
       "      <td>BiGG_core</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Population  Gene  locus_tag blattner  gene_length  \\\n",
       "0       Ara-1  appB  ECB_00982    b0979         1137   \n",
       "1       Ara-1  pykF  ECB_01645    b1676         1413   \n",
       "2       Ara-1  gltB  ECB_03077    b3212         4554   \n",
       "3       Ara-1  pflC  ECB_03837    b3952          879   \n",
       "4       Ara-2  pfkB  ECB_01692    b1723          930   \n",
       "5       Ara-2  fbaB  ECB_02025    b2097         1053   \n",
       "6       Ara-2   dld  ECB_02063    b2133         1716   \n",
       "7       Ara-2  eutI  ECB_02349    b2458         1017   \n",
       "8       Ara-3  kgtP  ECB_02481    b2587         1299   \n",
       "9       Ara-3  tdcE  ECB_02981    b3114         2295   \n",
       "10      Ara-3  tdcD  ECB_02982    b3115         1221   \n",
       "11      Ara-3  aceA  ECB_03887    b4015         1305   \n",
       "12      Ara-3  rpiB  ECB_03962    b4090          450   \n",
       "13      Ara-3  fumB  ECB_03993    b4122         1647   \n",
       "14      Ara-4  pflB  ECB_00907    b0903         2283   \n",
       "15      Ara-4  acnA  ECB_01253    b1276         2676   \n",
       "16      Ara-4  sfcA  ECB_01437    b1479         1698   \n",
       "17      Ara-4  fumC  ECB_01580    b1611         1404   \n",
       "18      Ara-4  fbaB  ECB_02025    b2097         1053   \n",
       "19      Ara-4  yghK  ECB_02843    b2975         1683   \n",
       "20      Ara-4  glcB  ECB_02844    b2976         2172   \n",
       "21      Ara-4  pitB  ECB_02863    b2987         1500   \n",
       "22      Ara-6  maeB  ECB_02354    b2463         2280   \n",
       "23      Ara-6  gltB  ECB_03077    b3212         4554   \n",
       "24      Ara-6  gltD  ECB_03078    b3213         1419   \n",
       "25      Ara+1  pflB  ECB_00907    b0903         2283   \n",
       "26      Ara+1  yghK  ECB_02843    b2975         1683   \n",
       "27      Ara+2  appC  ECB_00981    b0978         1545   \n",
       "28      Ara+2  tktB  ECB_02356    b2465         2004   \n",
       "29      Ara+2  sgcE  ECB_04166    b4301          633   \n",
       "30      Ara+3  pflB  ECB_00907    b0903         2283   \n",
       "31      Ara+3  adhE  ECB_01215    b1241         2676   \n",
       "32      Ara+3  focB  ECB_02384    b2492          849   \n",
       "33      Ara+4  pykF  ECB_01645    b1676         1413   \n",
       "34      Ara+4  fbaB  ECB_02025    b2097         1053   \n",
       "35      Ara+4  gltB  ECB_03077    b3212         4554   \n",
       "36      Ara+5  yghK  ECB_02843    b2975         1683   \n",
       "37      Ara+5  glcB  ECB_02844    b2976         2172   \n",
       "38      Ara+5  pitB  ECB_02863    b2987         1500   \n",
       "39      Ara+5  gltP  ECB_03949    b4077         1314   \n",
       "40      Ara+6  yghK  ECB_02843    b2975         1683   \n",
       "\n",
       "                                              product    start      end  \\\n",
       "0                 cytochrome bd-II oxidase subunit II  1056387  1057523   \n",
       "1                                     pyruvate kinase  1732965  1734377   \n",
       "2                    glutamate synthase large subunit  3289867  3294420   \n",
       "3                  pyruvate formate lyase II activase  4126365  4127243   \n",
       "4                            6-phosphofructokinase II  1783635  1784564   \n",
       "5                      fructose-bisphosphate aldolase  2131641  2132693   \n",
       "6   D-lactate dehydrogenase FAD-binding NADH indep...  2173423  2175138   \n",
       "7             predicted phosphotransacetylase subunit  2493318  2494334   \n",
       "8                     alpha-ketoglutarate transporter  2645941  2647239   \n",
       "9   pyruvate formate-lyase 4/2-ketobutyrate format...  3193158  3195452   \n",
       "10       propionate kinase/acetate kinase C anaerobic  3195486  3196706   \n",
       "11                                   isocitrate lyase  4196554  4197858   \n",
       "12                     ribose-5-phosphate isomerase B  4292172  4292621   \n",
       "13    anaerobic class I fumarate hydratase fumarase B  4324395  4326041   \n",
       "14                           pyruvate formate lyase I   968365   970647   \n",
       "15                                aconitate hydratase  1334203  1336878   \n",
       "16  malate dehydrogenase decarboxylating NAD-requi...  1527910  1529607   \n",
       "17                                 fumarate hydratase  1662380  1663783   \n",
       "18                     fructose-bisphosphate aldolase  2131641  2132693   \n",
       "19                              glycolate transporter  3046489  3048171   \n",
       "20                                    malate synthase  3048526  3050697   \n",
       "21                              phosphate transporter  3068959  3070458   \n",
       "22                                       malic enzyme  2496926  2499205   \n",
       "23                   glutamate synthase large subunit  3289867  3294420   \n",
       "24    glutamate synthase 4Fe-4S protein small subunit  3294433  3295851   \n",
       "25                           pyruvate formate lyase I   968365   970647   \n",
       "26                              glycolate transporter  3046489  3048171   \n",
       "27                 cytochrome bd-II oxidase subunit I  1054831  1056375   \n",
       "28                    transketolase 2 thiamin-binding  2500464  2502467   \n",
       "29                     ribulose-phosphate 3-epimerase  4507043  4507675   \n",
       "30                           pyruvate formate lyase I   968365   970647   \n",
       "31  fused acetaldehyde-CoA dehydrogenase/iron-depe...  1294239  1296914   \n",
       "32                      predicted formate transporter  2534754  2535602   \n",
       "33                                    pyruvate kinase  1732965  1734377   \n",
       "34                     fructose-bisphosphate aldolase  2131641  2132693   \n",
       "35                   glutamate synthase large subunit  3289867  3294420   \n",
       "36                              glycolate transporter  3046489  3048171   \n",
       "37                                    malate synthase  3048526  3050697   \n",
       "38                              phosphate transporter  3068959  3070458   \n",
       "39               glutamate/aspartate:proton symporter  4273414  4274727   \n",
       "40                              glycolate transporter  3046489  3048171   \n",
       "\n",
       "    strand MetabolicClass  \n",
       "0        1      BiGG_core  \n",
       "1        1      BiGG_core  \n",
       "2        1      BiGG_core  \n",
       "3        1      BiGG_core  \n",
       "4        1      BiGG_core  \n",
       "5       -1      BiGG_core  \n",
       "6        1      BiGG_core  \n",
       "7       -1      BiGG_core  \n",
       "8       -1      BiGG_core  \n",
       "9       -1      BiGG_core  \n",
       "10      -1      BiGG_core  \n",
       "11       1      BiGG_core  \n",
       "12       1      BiGG_core  \n",
       "13      -1      BiGG_core  \n",
       "14      -1      BiGG_core  \n",
       "15       1      BiGG_core  \n",
       "16      -1      BiGG_core  \n",
       "17      -1      BiGG_core  \n",
       "18      -1      BiGG_core  \n",
       "19      -1      BiGG_core  \n",
       "20      -1      BiGG_core  \n",
       "21      -1      BiGG_core  \n",
       "22      -1      BiGG_core  \n",
       "23       1      BiGG_core  \n",
       "24       1      BiGG_core  \n",
       "25      -1      BiGG_core  \n",
       "26      -1      BiGG_core  \n",
       "27       1      BiGG_core  \n",
       "28       1      BiGG_core  \n",
       "29      -1      BiGG_core  \n",
       "30      -1      BiGG_core  \n",
       "31      -1      BiGG_core  \n",
       "32       1      BiGG_core  \n",
       "33       1      BiGG_core  \n",
       "34      -1      BiGG_core  \n",
       "35       1      BiGG_core  \n",
       "36      -1      BiGG_core  \n",
       "37      -1      BiGG_core  \n",
       "38      -1      BiGG_core  \n",
       "39       1      BiGG_core  \n",
       "40      -1      BiGG_core  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KOed_BiGG_core"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Gene</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>blattner</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>product</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>MetabolicClass</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>pgpB</td>\n",
       "      <td>ECB_01255</td>\n",
       "      <td>b1278</td>\n",
       "      <td>765</td>\n",
       "      <td>phosphatidylglycerophosphatase B</td>\n",
       "      <td>1337702</td>\n",
       "      <td>1338466</td>\n",
       "      <td>1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>mrcB</td>\n",
       "      <td>ECB_00148</td>\n",
       "      <td>b0149</td>\n",
       "      <td>2526</td>\n",
       "      <td>penicillin-binding protein 1b</td>\n",
       "      <td>167581</td>\n",
       "      <td>170106</td>\n",
       "      <td>1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>Ara-2</td>\n",
       "      <td>thiD</td>\n",
       "      <td>ECB_02031</td>\n",
       "      <td>b2103</td>\n",
       "      <td>801</td>\n",
       "      <td>phosphomethylpyrimidine kinase</td>\n",
       "      <td>2137845</td>\n",
       "      <td>2138645</td>\n",
       "      <td>-1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>Ara-4</td>\n",
       "      <td>thiD</td>\n",
       "      <td>ECB_02031</td>\n",
       "      <td>b2103</td>\n",
       "      <td>801</td>\n",
       "      <td>phosphomethylpyrimidine kinase</td>\n",
       "      <td>2137845</td>\n",
       "      <td>2138645</td>\n",
       "      <td>-1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>Ara-5</td>\n",
       "      <td>ispF</td>\n",
       "      <td>ECB_02596</td>\n",
       "      <td>b2746</td>\n",
       "      <td>480</td>\n",
       "      <td>2-C-methyl-D-erythritol 24-cyclodiphosphate sy...</td>\n",
       "      <td>2766162</td>\n",
       "      <td>2766641</td>\n",
       "      <td>-1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>Ara-5</td>\n",
       "      <td>ispD</td>\n",
       "      <td>ECB_02597</td>\n",
       "      <td>b2747</td>\n",
       "      <td>711</td>\n",
       "      <td>2-C-methyl-D-erythritol 4-phosphate cytidylylt...</td>\n",
       "      <td>2766641</td>\n",
       "      <td>2767351</td>\n",
       "      <td>-1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>mrdA</td>\n",
       "      <td>ECB_00604</td>\n",
       "      <td>b0635</td>\n",
       "      <td>1902</td>\n",
       "      <td>transpeptidase involved in peptidoglycan synth...</td>\n",
       "      <td>648900</td>\n",
       "      <td>650801</td>\n",
       "      <td>-1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>ispF</td>\n",
       "      <td>ECB_02596</td>\n",
       "      <td>b2746</td>\n",
       "      <td>480</td>\n",
       "      <td>2-C-methyl-D-erythritol 24-cyclodiphosphate sy...</td>\n",
       "      <td>2766162</td>\n",
       "      <td>2766641</td>\n",
       "      <td>-1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>ispD</td>\n",
       "      <td>ECB_02597</td>\n",
       "      <td>b2747</td>\n",
       "      <td>711</td>\n",
       "      <td>2-C-methyl-D-erythritol 4-phosphate cytidylylt...</td>\n",
       "      <td>2766641</td>\n",
       "      <td>2767351</td>\n",
       "      <td>-1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>Ara+3</td>\n",
       "      <td>thiG</td>\n",
       "      <td>ECB_03867</td>\n",
       "      <td>b3991</td>\n",
       "      <td>771</td>\n",
       "      <td>thiazole synthase</td>\n",
       "      <td>4171476</td>\n",
       "      <td>4172246</td>\n",
       "      <td>-1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>Ara+4</td>\n",
       "      <td>thiD</td>\n",
       "      <td>ECB_02031</td>\n",
       "      <td>b2103</td>\n",
       "      <td>801</td>\n",
       "      <td>phosphomethylpyrimidine kinase</td>\n",
       "      <td>2137845</td>\n",
       "      <td>2138645</td>\n",
       "      <td>-1</td>\n",
       "      <td>Superessential</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Population  Gene  locus_tag blattner  gene_length  \\\n",
       "41      Ara-1  pgpB  ECB_01255    b1278          765   \n",
       "42      Ara-2  mrcB  ECB_00148    b0149         2526   \n",
       "43      Ara-2  thiD  ECB_02031    b2103          801   \n",
       "44      Ara-4  thiD  ECB_02031    b2103          801   \n",
       "45      Ara-5  ispF  ECB_02596    b2746          480   \n",
       "46      Ara-5  ispD  ECB_02597    b2747          711   \n",
       "47      Ara+1  mrdA  ECB_00604    b0635         1902   \n",
       "48      Ara+1  ispF  ECB_02596    b2746          480   \n",
       "49      Ara+1  ispD  ECB_02597    b2747          711   \n",
       "50      Ara+3  thiG  ECB_03867    b3991          771   \n",
       "51      Ara+4  thiD  ECB_02031    b2103          801   \n",
       "\n",
       "                                              product    start      end  \\\n",
       "41                   phosphatidylglycerophosphatase B  1337702  1338466   \n",
       "42                      penicillin-binding protein 1b   167581   170106   \n",
       "43                     phosphomethylpyrimidine kinase  2137845  2138645   \n",
       "44                     phosphomethylpyrimidine kinase  2137845  2138645   \n",
       "45  2-C-methyl-D-erythritol 24-cyclodiphosphate sy...  2766162  2766641   \n",
       "46  2-C-methyl-D-erythritol 4-phosphate cytidylylt...  2766641  2767351   \n",
       "47  transpeptidase involved in peptidoglycan synth...   648900   650801   \n",
       "48  2-C-methyl-D-erythritol 24-cyclodiphosphate sy...  2766162  2766641   \n",
       "49  2-C-methyl-D-erythritol 4-phosphate cytidylylt...  2766641  2767351   \n",
       "50                                  thiazole synthase  4171476  4172246   \n",
       "51                     phosphomethylpyrimidine kinase  2137845  2138645   \n",
       "\n",
       "    strand  MetabolicClass  \n",
       "41       1  Superessential  \n",
       "42       1  Superessential  \n",
       "43      -1  Superessential  \n",
       "44      -1  Superessential  \n",
       "45      -1  Superessential  \n",
       "46      -1  Superessential  \n",
       "47      -1  Superessential  \n",
       "48      -1  Superessential  \n",
       "49      -1  Superessential  \n",
       "50      -1  Superessential  \n",
       "51      -1  Superessential  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "KOed_superessential"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic metabolic models to play with. Default objective is to maximize biomass."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "BiGG_model_dir = \"../data/BiGG-models\"\n",
    "\n",
    "# the E. coli K-12 iJO1366 model: this is the best curated complete E. coli model.\n",
    "K12_model = cobra.io.load_json_model(path.join(BiGG_model_dir, \"iJO1366.json\"))\n",
    "K12_model.id = \"K12\"\n",
    "\n",
    "# the E. coli REL606 iECB_1328 model: most relevant to LTEE, but has stochiometric inconsistencies\n",
    "## in hydrogen/proton conservation.\n",
    "REL606_model = cobra.io.load_json_model(path.join(BiGG_model_dir, \"iECB_1328.json\"))\n",
    "REL606_model.id = \"REL606\"\n",
    "\n",
    "## The simplest well-curated model: E. coli core metabolism.\n",
    "core_model = cobra.io.load_json_model(path.join(BiGG_model_dir, \"e_coli_core.json\"))\n",
    "core_model.id = \"Ecoli-core\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_glc__D_e': 10.0, 'EX_h_e': 1000.0, 'EX_h2o_e': 1000.0, 'EX_k_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_nh4_e': 1000.0, 'EX_ca2_e': 1000.0, 'EX_cbl1_e': 0.01, 'EX_ni2_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_cl_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_zn2_e': 1000.0, 'EX_sel_e': 1000.0, 'EX_slnt_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_fe3_e': 1000.0}\n",
      "{'EX_ca2_e': 1000.0, 'EX_cbl1_e': 0.01, 'EX_cl_e': 1000.0, 'EX_co2_e': 1000.0, 'EX_cobalt2_e': 1000.0, 'EX_cu2_e': 1000.0, 'EX_fe2_e': 1000.0, 'EX_fe3_e': 1000.0, 'EX_mg2_e': 1000.0, 'EX_mn2_e': 1000.0, 'EX_mobd_e': 1000.0, 'EX_na1_e': 1000.0, 'EX_nh4_e': 1000.0, 'EX_ni2_e': 1000.0, 'EX_glc__D_e': 10.0, 'EX_o2_e': 1000.0, 'EX_tungs_e': 1000.0, 'EX_pi_e': 1000.0, 'EX_zn2_e': 1000.0, 'EX_sel_e': 1000.0, 'EX_slnt_e': 1000.0, 'EX_so4_e': 1000.0, 'EX_h_e': 1000.0, 'EX_h2o_e': 1000.0, 'EX_k_e': 1000.0}\n",
      "{'EX_co2_e': 1000.0, 'EX_glc__D_e': 10.0, 'EX_h_e': 1000.0, 'EX_h2o_e': 1000.0, 'EX_nh4_e': 1000.0, 'EX_o2_e': 1000.0, 'EX_pi_e': 1000.0}\n"
     ]
    }
   ],
   "source": [
    "## we are simulating media in which glucose import is limiting.\n",
    "print(K12_model.medium)\n",
    "print(REL606_model.medium)\n",
    "print(core_model.medium)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 1: Measure fitness of 50K A clones using just the E. coli core model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "core_LTEE_models = generate_LTEE_clone_cobra_models(KOed_BiGG_core, core_model)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Measure the single-KO DFE for each 50K A clone, just using the core metabolic model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "nonmutator_pops = [\"Ara-5\", \"Ara-6\", \"Ara+1\", \"Ara+2\", \"Ara+4\", \"Ara+5\"]\n",
    "mutator_pops = [\"Ara-1\", \"Ara-2\", \"Ara-3\", \"Ara-4\", \"Ara+3\", \"Ara+6\"]\n",
    "LTEE_pop_vec = nonmutator_pops + mutator_pops"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 2: Measure fitness of 50K A clones using the whole genome metabolic model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Gene</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>blattner</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>product</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>mutT</td>\n",
       "      <td>ECB_00100</td>\n",
       "      <td>b0099</td>\n",
       "      <td>390</td>\n",
       "      <td>nucleoside triphosphate pyrophosphohydrolase m...</td>\n",
       "      <td>113848</td>\n",
       "      <td>114237</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>ybaL</td>\n",
       "      <td>ECB_00429</td>\n",
       "      <td>b0478</td>\n",
       "      <td>1677</td>\n",
       "      <td>predicted transporter with NADP-binding Rossma...</td>\n",
       "      <td>473629</td>\n",
       "      <td>475305</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>ybcR</td>\n",
       "      <td>ECB_00506</td>\n",
       "      <td>b0554</td>\n",
       "      <td>216</td>\n",
       "      <td>predicted phage lysis protein</td>\n",
       "      <td>548867</td>\n",
       "      <td>549082</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>ybcS</td>\n",
       "      <td>ECB_00507</td>\n",
       "      <td>b0555</td>\n",
       "      <td>498</td>\n",
       "      <td>predicted lysozyme</td>\n",
       "      <td>549082</td>\n",
       "      <td>549579</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Ara-1</td>\n",
       "      <td>ybcT</td>\n",
       "      <td>ECB_00508</td>\n",
       "      <td>b0556</td>\n",
       "      <td>462</td>\n",
       "      <td>predicted murein endopeptidase</td>\n",
       "      <td>549576</td>\n",
       "      <td>550037</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2290</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>rbsR</td>\n",
       "      <td>ECB_03639</td>\n",
       "      <td>b3753</td>\n",
       "      <td>993</td>\n",
       "      <td>DNA-binding transcriptional repressor of ribos...</td>\n",
       "      <td>3900034</td>\n",
       "      <td>3901026</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2291</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>frvR</td>\n",
       "      <td>ECB_03782</td>\n",
       "      <td>b3897</td>\n",
       "      <td>1749</td>\n",
       "      <td>predicted regulator</td>\n",
       "      <td>4066254</td>\n",
       "      <td>4068002</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2292</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>metH</td>\n",
       "      <td>ECB_03891</td>\n",
       "      <td>b4019</td>\n",
       "      <td>3684</td>\n",
       "      <td>B12-dependent methionine synthase</td>\n",
       "      <td>4202759</td>\n",
       "      <td>4206442</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2293</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>sgcQ</td>\n",
       "      <td>ECB_04168</td>\n",
       "      <td>b4303</td>\n",
       "      <td>807</td>\n",
       "      <td>predicted nucleoside triphosphatase</td>\n",
       "      <td>4508249</td>\n",
       "      <td>4509055</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2294</th>\n",
       "      <td>Ara+6</td>\n",
       "      <td>smp</td>\n",
       "      <td>ECB_04263</td>\n",
       "      <td>b4387</td>\n",
       "      <td>645</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>4612359</td>\n",
       "      <td>4613003</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2295 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Population  Gene  locus_tag blattner  gene_length  \\\n",
       "0         Ara-1  mutT  ECB_00100    b0099          390   \n",
       "1         Ara-1  ybaL  ECB_00429    b0478         1677   \n",
       "2         Ara-1  ybcR  ECB_00506    b0554          216   \n",
       "3         Ara-1  ybcS  ECB_00507    b0555          498   \n",
       "4         Ara-1  ybcT  ECB_00508    b0556          462   \n",
       "...         ...   ...        ...      ...          ...   \n",
       "2290      Ara+6  rbsR  ECB_03639    b3753          993   \n",
       "2291      Ara+6  frvR  ECB_03782    b3897         1749   \n",
       "2292      Ara+6  metH  ECB_03891    b4019         3684   \n",
       "2293      Ara+6  sgcQ  ECB_04168    b4303          807   \n",
       "2294      Ara+6   smp  ECB_04263    b4387          645   \n",
       "\n",
       "                                                product    start      end  \\\n",
       "0     nucleoside triphosphate pyrophosphohydrolase m...   113848   114237   \n",
       "1     predicted transporter with NADP-binding Rossma...   473629   475305   \n",
       "2                         predicted phage lysis protein   548867   549082   \n",
       "3                                    predicted lysozyme   549082   549579   \n",
       "4                        predicted murein endopeptidase   549576   550037   \n",
       "...                                                 ...      ...      ...   \n",
       "2290  DNA-binding transcriptional repressor of ribos...  3900034  3901026   \n",
       "2291                                predicted regulator  4066254  4068002   \n",
       "2292                  B12-dependent methionine synthase  4202759  4206442   \n",
       "2293                predicted nucleoside triphosphatase  4508249  4509055   \n",
       "2294                               hypothetical protein  4612359  4613003   \n",
       "\n",
       "      strand  \n",
       "0          1  \n",
       "1         -1  \n",
       "2          1  \n",
       "3          1  \n",
       "4          1  \n",
       "...      ...  \n",
       "2290       1  \n",
       "2291      -1  \n",
       "2292       1  \n",
       "2293      -1  \n",
       "2294      -1  \n",
       "\n",
       "[2295 rows x 9 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_KOed_genes_in_50K_A_clones = pd.read_csv(\"../results/metabolic-enzymes/KOed-genes-in-LTEE-50K-A-clones.csv\")\n",
    "all_KOed_genes_in_50K_A_clones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population</th>\n",
       "      <th>Gene</th>\n",
       "      <th>locus_tag</th>\n",
       "      <th>blattner</th>\n",
       "      <th>gene_length</th>\n",
       "      <th>product</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1194</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>nhaA</td>\n",
       "      <td>ECB_00018</td>\n",
       "      <td>b0019</td>\n",
       "      <td>1167</td>\n",
       "      <td>pH-dependent sodium/proton antiporter</td>\n",
       "      <td>17488</td>\n",
       "      <td>18654</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>dgt</td>\n",
       "      <td>ECB_00159</td>\n",
       "      <td>b0160</td>\n",
       "      <td>1518</td>\n",
       "      <td>deoxyguanosinetriphosphate triphosphohydrolase</td>\n",
       "      <td>182079</td>\n",
       "      <td>183596</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1196</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>lacZ</td>\n",
       "      <td>ECB_00298</td>\n",
       "      <td>b0344</td>\n",
       "      <td>3075</td>\n",
       "      <td>beta-D-galactosidase</td>\n",
       "      <td>335787</td>\n",
       "      <td>338861</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1197</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>ybbO</td>\n",
       "      <td>ECB_00444</td>\n",
       "      <td>b0493</td>\n",
       "      <td>810</td>\n",
       "      <td>short chain dehydrogenase</td>\n",
       "      <td>490508</td>\n",
       "      <td>491317</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1198</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>tesA</td>\n",
       "      <td>ECB_00445</td>\n",
       "      <td>b0494</td>\n",
       "      <td>627</td>\n",
       "      <td>multifunctional acyl-CoA thioesterase I and pr...</td>\n",
       "      <td>491307</td>\n",
       "      <td>491933</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1517</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>yjcO</td>\n",
       "      <td>ECB_03950</td>\n",
       "      <td>b4078</td>\n",
       "      <td>690</td>\n",
       "      <td>hypothetical protein</td>\n",
       "      <td>4275258</td>\n",
       "      <td>4275947</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>rpiR</td>\n",
       "      <td>ECB_03961</td>\n",
       "      <td>b4089</td>\n",
       "      <td>891</td>\n",
       "      <td>DNA-binding transcriptional repressor</td>\n",
       "      <td>4290923</td>\n",
       "      <td>4291813</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1519</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>cycA</td>\n",
       "      <td>ECB_04080</td>\n",
       "      <td>b4208</td>\n",
       "      <td>1413</td>\n",
       "      <td>D-alanine/D-serine/glycine transporter</td>\n",
       "      <td>4414863</td>\n",
       "      <td>4416275</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>hsdS</td>\n",
       "      <td>ECB_04214</td>\n",
       "      <td>b4348</td>\n",
       "      <td>1425</td>\n",
       "      <td>specificity determinant for hsdM and hsdR</td>\n",
       "      <td>4558013</td>\n",
       "      <td>4559437</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1521</th>\n",
       "      <td>Ara+1</td>\n",
       "      <td>nadR</td>\n",
       "      <td>ECB_04266</td>\n",
       "      <td>b4390</td>\n",
       "      <td>1233</td>\n",
       "      <td>nicotinamide-nucleotide adenylyltransferase</td>\n",
       "      <td>4615529</td>\n",
       "      <td>4616761</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>328 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Population  Gene  locus_tag blattner  gene_length  \\\n",
       "1194      Ara+1  nhaA  ECB_00018    b0019         1167   \n",
       "1195      Ara+1   dgt  ECB_00159    b0160         1518   \n",
       "1196      Ara+1  lacZ  ECB_00298    b0344         3075   \n",
       "1197      Ara+1  ybbO  ECB_00444    b0493          810   \n",
       "1198      Ara+1  tesA  ECB_00445    b0494          627   \n",
       "...         ...   ...        ...      ...          ...   \n",
       "1517      Ara+1  yjcO  ECB_03950    b4078          690   \n",
       "1518      Ara+1  rpiR  ECB_03961    b4089          891   \n",
       "1519      Ara+1  cycA  ECB_04080    b4208         1413   \n",
       "1520      Ara+1  hsdS  ECB_04214    b4348         1425   \n",
       "1521      Ara+1  nadR  ECB_04266    b4390         1233   \n",
       "\n",
       "                                                product    start      end  \\\n",
       "1194              pH-dependent sodium/proton antiporter    17488    18654   \n",
       "1195     deoxyguanosinetriphosphate triphosphohydrolase   182079   183596   \n",
       "1196                               beta-D-galactosidase   335787   338861   \n",
       "1197                          short chain dehydrogenase   490508   491317   \n",
       "1198  multifunctional acyl-CoA thioesterase I and pr...   491307   491933   \n",
       "...                                                 ...      ...      ...   \n",
       "1517                               hypothetical protein  4275258  4275947   \n",
       "1518              DNA-binding transcriptional repressor  4290923  4291813   \n",
       "1519             D-alanine/D-serine/glycine transporter  4414863  4416275   \n",
       "1520          specificity determinant for hsdM and hsdR  4558013  4559437   \n",
       "1521        nicotinamide-nucleotide adenylyltransferase  4615529  4616761   \n",
       "\n",
       "      strand  \n",
       "1194       1  \n",
       "1195       1  \n",
       "1196      -1  \n",
       "1197      -1  \n",
       "1198      -1  \n",
       "...      ...  \n",
       "1517      -1  \n",
       "1518      -1  \n",
       "1519       1  \n",
       "1520      -1  \n",
       "1521       1  \n",
       "\n",
       "[328 rows x 9 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_KOed_genes_in_50K_A_clones[all_KOed_genes_in_50K_A_clones.Population == \"Ara+1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Generate models from K-12.\n",
    "iJO1366_LTEE_models = generate_LTEE_clone_cobra_models(all_KOed_genes_in_50K_A_clones, K12_model)\n",
    "## Generate models from REL606.\n",
    "iECB_1328_LTEE_models = generate_LTEE_clone_cobra_models(all_KOed_genes_in_50K_A_clones, REL606_model, using_locus_tag=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 3: Examine knockouts of BiGG core genes and superessential genes in the 50K genomes.\n",
    "## Is there variation in the DFE across LTEE populations?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "##BiGG_core_genes = pd.read_csv(\"\")\n",
    "##bigg_core_gene_vec = "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "superessential_genes_df = pd.read_csv(\"../results/metabolic-enzymes/Barve2012-S6-superessential.csv\")\n",
    "superessential_gene_vec = [x for x in superessential_genes_df.blattner]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Question 4: Generate minimal metabolic networks.\n",
    "\n",
    "Question: How much variation in how long to reach a minimal genome?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def play_jenga(basic_model, cutoff = 0.01):\n",
    "    \"\"\"\n",
    "    While there exist genes that have a beneficial effect when knocked out:\n",
    "    - randomly order the genes to remove.\n",
    "    - generate a KO. If deleterious, then reject the change and continue.\n",
    "      If the change is beneficial, then accept and update the model.\n",
    "    - once the entire DFE of possible KOs is deleterious,\n",
    "    return the evolved model.\n",
    "    \"\"\"\n",
    "    random.seed() ## seed the PRNG with the current system time.\n",
    "    evolved_model = basic_model.copy()\n",
    "    evolved_model.id = \"jenga_evolved\"\n",
    "    \n",
    "    steps = 0 ## number of genes KO'ed so far\n",
    "    while(True):\n",
    "        cur_biomass = evolved_model.slim_optimize()\n",
    "        KO_results = cobra.flux_analysis.single_gene_deletion(evolved_model)\n",
    "        num_genes = len(KO_results.ids)\n",
    "        print(\"num_genes\", num_genes)\n",
    "        \n",
    "        \"\"\" From the Methods of Pal et al. (2006) in Nature:\n",
    "        A gene was classified as having no fitness effect \n",
    "        if the biomass production rate of the knockout strain \n",
    "        was reduced by less than a given cutoff; different cutoffs\n",
    "        led to very similar results.\n",
    "        \n",
    "        filter potential knockouts based on this criterion.\n",
    "        Pal et al. used cutoff = 0.01 and cutoff = 0.1.\"\"\"\n",
    "\n",
    "        filtered_KO_results = KO_results[KO_results.growth > (cur_biomass - cutoff)]\n",
    "        num_candidate_genes = len(filtered_KO_results.ids)\n",
    "        fract_purifying = float(num_genes - num_candidate_genes)/float(num_genes)\n",
    "        print(\"num_candidate_genes\", num_candidate_genes)\n",
    "        print(\"fraction genes under purifying selection: \", fract_purifying)\n",
    "        if num_candidate_genes: ## there exist genes to delete\n",
    "            idx_to_delete = random.sample(range(num_candidate_genes),1).pop()\n",
    "            genes_to_delete = filtered_KO_results.ids.tolist()[idx_to_delete]\n",
    "            print(\"deleting: \", genes_to_delete)\n",
    "            steps += 1\n",
    "            print(\"steps <= \" + str(steps))\n",
    "            cobra.manipulation.remove_genes(evolved_model, genes_to_delete)\n",
    "        else: ## no more genes to delete\n",
    "            break ## leave the while loop.  \n",
    "    print(\"steps <= \" + str(steps))\n",
    "    return evolved_model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "num_genes 1329\n",
      "num_candidate_genes 1072\n",
      "fraction genes under purifying selection:  0.19337848006019565\n",
      "deleting:  {'ECB_01632'}\n",
      "steps <= 1\n",
      "num_genes 1328\n",
      "num_candidate_genes 1071\n",
      "fraction genes under purifying selection:  0.19352409638554216\n",
      "deleting:  {'ECB_01401'}\n",
      "steps <= 2\n",
      "num_genes 1327\n",
      "num_candidate_genes 1070\n",
      "fraction genes under purifying selection:  0.19366993217784476\n",
      "deleting:  {'ECB_02059'}\n",
      "steps <= 3\n",
      "num_genes 1326\n",
      "num_candidate_genes 1069\n",
      "fraction genes under purifying selection:  0.193815987933635\n",
      "deleting:  {'ECB_03478'}\n",
      "steps <= 4\n",
      "num_genes 1325\n",
      "num_candidate_genes 1068\n",
      "fraction genes under purifying selection:  0.1939622641509434\n",
      "deleting:  {'ECB_02574'}\n",
      "steps <= 5\n",
      "num_genes 1324\n",
      "num_candidate_genes 1067\n",
      "fraction genes under purifying selection:  0.19410876132930513\n",
      "deleting:  {'ECB_00556'}\n",
      "steps <= 6\n",
      "num_genes 1323\n",
      "num_candidate_genes 1066\n",
      "fraction genes under purifying selection:  0.1942554799697657\n",
      "deleting:  {'ECB_00687'}\n",
      "steps <= 7\n",
      "num_genes 1322\n",
      "num_candidate_genes 1060\n",
      "fraction genes under purifying selection:  0.19818456883509833\n",
      "deleting:  {'ECB_03662'}\n",
      "steps <= 8\n",
      "num_genes 1321\n",
      "num_candidate_genes 1059\n",
      "fraction genes under purifying selection:  0.198334595003785\n",
      "deleting:  {'ECB_01837'}\n",
      "steps <= 9\n",
      "num_genes 1320\n",
      "num_candidate_genes 1058\n",
      "fraction genes under purifying selection:  0.1984848484848485\n",
      "deleting:  {'ECB_02079'}\n",
      "steps <= 10\n",
      "num_genes 1319\n",
      "num_candidate_genes 1057\n",
      "fraction genes under purifying selection:  0.19863532979529946\n",
      "deleting:  {'ECB_01204'}\n",
      "steps <= 11\n",
      "num_genes 1318\n",
      "num_candidate_genes 1056\n",
      "fraction genes under purifying selection:  0.19878603945371776\n",
      "deleting:  {'ECB_01645'}\n",
      "steps <= 12\n",
      "num_genes 1317\n",
      "num_candidate_genes 1055\n",
      "fraction genes under purifying selection:  0.19893697798025817\n",
      "deleting:  {'ECB_00359'}\n",
      "steps <= 13\n",
      "num_genes 1316\n",
      "num_candidate_genes 1054\n",
      "fraction genes under purifying selection:  0.19908814589665655\n",
      "deleting:  {'ECB_04025'}\n",
      "steps <= 14\n",
      "num_genes 1315\n",
      "num_candidate_genes 1053\n",
      "fraction genes under purifying selection:  0.19923954372623573\n",
      "deleting:  {'ECB_01728'}\n",
      "steps <= 15\n",
      "num_genes 1314\n",
      "num_candidate_genes 1052\n",
      "fraction genes under purifying selection:  0.1993911719939117\n",
      "deleting:  {'ECB_04258'}\n",
      "steps <= 16\n",
      "num_genes 1313\n",
      "num_candidate_genes 1051\n",
      "fraction genes under purifying selection:  0.19954303122619954\n",
      "deleting:  {'ECB_00961'}\n",
      "steps <= 17\n",
      "num_genes 1312\n",
      "num_candidate_genes 1050\n",
      "fraction genes under purifying selection:  0.19969512195121952\n",
      "deleting:  {'ECB_00580'}\n",
      "steps <= 18\n",
      "num_genes 1311\n",
      "num_candidate_genes 1049\n",
      "fraction genes under purifying selection:  0.19984744469870327\n",
      "deleting:  {'ECB_00941'}\n",
      "steps <= 19\n",
      "num_genes 1310\n",
      "num_candidate_genes 1048\n",
      "fraction genes under purifying selection:  0.2\n",
      "deleting:  {'ECB_02420'}\n",
      "steps <= 20\n",
      "num_genes 1309\n",
      "num_candidate_genes 1041\n",
      "fraction genes under purifying selection:  0.2047364400305577\n",
      "deleting:  {'ECB_02962'}\n",
      "steps <= 21\n",
      "num_genes 1308\n",
      "num_candidate_genes 1040\n",
      "fraction genes under purifying selection:  0.20489296636085627\n",
      "deleting:  {'ECB_00198'}\n",
      "steps <= 22\n",
      "num_genes 1307\n",
      "num_candidate_genes 1039\n",
      "fraction genes under purifying selection:  0.20504973221117062\n",
      "deleting:  {'ECB_00037'}\n",
      "steps <= 23\n",
      "num_genes 1306\n",
      "num_candidate_genes 1038\n",
      "fraction genes under purifying selection:  0.20520673813169985\n",
      "deleting:  {'ECB_01275'}\n",
      "steps <= 24\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "evolved_model = play_jenga(REL606_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## 1) generate minimal genomes (see Pal et al. 2006 in Nature, referenced by Hosseini et al. 2018 in PNAS).\n",
    "##    while generating minimal genomes, plot the percentage of essential genes in the network.\n",
    "##    Also examine the percentage of essential, superessential, and BiGG core, specialist, and generalist enzymes\n",
    "##    over time.\n",
    "## 2) show predicted viability of the evolved LTEE metabolic networks over time.\n",
    "\n",
    "evolved_model.optimize()\n",
    "len(evolved_model.genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(REL606_model.genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "137-71"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
